src/core/dataset.rs

```diff
@@ -125,13 +132,55 @@ impl<T: Number, U: Number> Dataset<T, U> for VecVec<T, U> {
     fn query_to_one(&self, query: &[T], index: usize) -> U {
         (self.metric)(query, &self.data[index])
     }
+
+    fn reorder(&mut self, indices: &[usize]) {
+        let n = indices.len();
+
+        // TODO: We'll need to support reordering only a subset (i.e. batch)
+        // of indices at some point, so this assert will change in the future.
+        assert!(n == self.cardinality());
+
+        // The "source index" represents the index that we hope to swap to
+        let mut source_index: usize;
+
+        // INVARIANT: After each iteration of the loop, the elements of the
+        // subarray [0..i] are in the correct position.
+        for i in 0..n-1 {
+            source_index = indices[i];
+
+            // If the element at is already at the correct position, we can
+            // just skip.
+            if source_index != i {
+                // Here we're essentially following the cycle. We *know* by
+                // the invariant that all elements to the left of i are in
+                // the correct position, so what we're doing is following
+                // the cycle until we find an index to the right of i. Which,
+                // because we followed the position changes, is the correct
+                // index to swap.
+                while source_index < i {
+                    source_index = indices[source_index];
+                }
+
+                // We swap to the correct index. Importantly, this index is always
+                // to the right of i, we do not modify any index to the left of i.
+                // Thus, because we followed the cycle to the correct index to swap,
+                // we know that the element at i, after this swap, is in the correct
+                // position.
+                self.swap(source_index, i);
+            }
+        }
+    }
```

Comment:
Move this implementation into the trait (at line 78), i.e. make it a default implementation.